src/colmap/scene/reconstruction.cc

```diff
@@ -262,11 +257,8 @@ void Reconstruction::DeRegisterImage(const image_t image_id) {
     }
   }
 
-  image.SetRegistered(false);
-
-  reg_image_ids_.erase(
-      std::remove(reg_image_ids_.begin(), reg_image_ids_.end(), image_id),
-      reg_image_ids_.end());
+  image.DeRegister();
```

Comment:
First I think the current logistics is a bit confusing: there are two levels of "registration", and they are not always consistent: 1) whether an image has a cam from world set 2) whether an image has been registered in the reconstruction with Reconstruction::RegisterImage.

Personally I don't think these two levels should be made equivalent, so would be better if we use a different word from "register" in the image class.

And above all, the image should anyway not be deregistered in Reconstruction::DeregisterImage. Only the removal from the reg_image_ids is needed here. Here exactly the confusion of different levels comes to life: with the current logic if we call Reconstruction::RegisterImage and then Reconstruction::DeRegisterImage, the image pose will be lost.

Reply:
This PR avoids breaking changes as much as possible. The fact that `Reconstruction::DeregisterImage` de-registers the image object as well is consistent with the current behavior and avoids silently breaking any user code. IMHO whether the image state should be called "registered" or something else is a question beyond this PR. Maybe @ahojnnes can weigh in there.

Reply:
Only with the change of this PR the image pose is lost after RegisterImage and DeRegisterImage, since before only the registered_ flag is changed. I think the registered_ flag before was conceptually different from whether the image has a pose or not in the design, but maybe I misunderstood here @ahojnnes 

P.S. (I agree that the following part might be beyond the scope of this PR) On the other hand, if the registered_ was meant to be equivalent to whether the image has a pose or not, it's a bit strange to have Reconstruction::RegisterImage and Reconstruction::DeregisterImage as all the operations should be on the image objects.

Reply:
I don't have a strong opinion but I would suggest removing the "registered" terminology completely from the image now and isolate it to the reconstruction, and simply have a "HasCamFromWorld()" for the image. This is me writing on the phone without being able to look at all usages, so this may not be an ideal suggestion.

Reply:
I have updated `{IsRegistered,DeRegister}` to `{Has,Reset}Pose` (less verbose than `CamFromWorld`). I have also updated `Reconstruction::IsImageRegistered` to use `reg_image_ids_`.  One last point: I still think that removing `Image::ResetPose` from `Reconstruction::DeRegisterImage` is a bad idea since we never do one without the other and it might silently break user code.

Reply:
Since we have completely disentangled `ResetPose` and `DeRegisterImage` in the new design, I don't think `ResetPose` is needed here. With the new optional design `ResetPose` will also lose the pose information while the previous code will not, so we will anyway need some efforts from the users if they hit this.